torch/autograd/gradcheck.py

```diff
@@ -72,6 +72,55 @@ def _iter_tensors(x: Union[torch.Tensor, Iterable[torch.Tensor]],
                 yield result
 
 
+def _densify(x):
+    # return a copy of sparse x with all unspecified elements
+    # "replaced" with zero-valued elements
+    if isinstance(x, (list, tuple)):
+        return type(x)(map(_densify, x))
+    elif not is_tensor_like(x) or x.layout in {torch.strided, torch._mkldnn}:  # type: ignore[attr-defined] # no attr _mkldnn
+        return x
+    elif x.layout is torch.sparse_coo:
+        def get_stride(size):
+            dim = len(size)
+            tmp = 1
+            stride = [0] * dim
+            for i in reversed(range(dim)):
+                stride[i] = tmp
+                tmp *= size[i]
+            return stride
+        stride = get_stride(x.shape[:x.sparse_dim()])
+        r = torch.ones(x.shape, dtype=x.dtype, device=x.device).to_sparse(
+            layout=x.layout, dense_dim=x.dense_dim())._coalesced_(True)
+        r._values().sub_(1)
+        # r is a sparse tensor with all explicit zeros
+        if 0 in x.shape:
+            return r.requires_grad_(x.requires_grad)
+        x = x.coalesce()
+        indices = x.indices().clone()
+        flat_indices = indices.mul_(torch.tensor(stride, device=indices.device, dtype=indices.dtype).unsqueeze(1)).sum(0)
+        r._values()[flat_indices] = x.values()
+        # r is now a sparse tensor with all elements specified and having values of x
```

Comment:
Could you explain why you do all these tricks to get here instead of just calling `torch.sparse_coo_tensor` with the constructed values and indices?

Reply:
Sure. To keep the story short, now values and indices are precomputed and the result is constructed using `sparse_coo_tensor`.